### PR TITLE
[Annotations] Introduce code-generated annotations and associated infrastructure.

### PR DESCRIPTION
**Pull request by nishant-karajgikar**
PRs must be submitted under the terms of our Contributor License Agreement [CLA](https://github.com/mapbox/mapbox-maps-ios/blob/main/CONTRIBUTING.md#contributor-license-agreement).

Fixes: < Link to related issues that will be fixed by this pull request, if they exist >

## Pull request checklist:
 - [x] Briefly describe the changes in this PR.
 - [ ] Include before/after visuals or gifs if this PR includes visual changes.
 - [x] Write tests for all new functionality. If tests were not written, please explain why.
 - [ ] Add example if relevant.
 - [x] Document any changes to public APIs.
 - [x] Apply changelog label ('breaking change', 'bug :beetle:', 'build', 'docs', 'feature :green_apple:', 'performance :zap:', 'testing :100:') or use the label 'skip changelog'
 - [ ] Add an entry inside this element for inclusion in the `mapbox-maps-ios` changelog: `<changelog></changelog>`.
 - [ ] Update the migration guide, API Docs, Markdown files - Readme, Developing, etc

### Summary of changes

This PR introduces a new approach to annotations, by creating object-oriented wrappers around sources/layers. 

#### MapView
- Renames original `annotations` property to `annotations_legacy`.
- Introduces new `annotations` property of type `AnnotationOrchestrator`.

#### AnnotationsOrchestrator
- Internal infrastructural object that exposes mechanisms to create various per-type annotation managers.
- Represents the entry point into annotations

#### Per-type annotation manager
- 4 types: `PointAnnotationManager`, `PolylineAnnotationManager`, `PolygonAnnotationManager`, `CircleAnnotationManager`.
- Each `*AnnotationManager` conforms to `AnnotationManager` protocol.
- Each `*AnnotationManager` is responsible for a collection (array) of its corresponding annotation.
- Every `*AnnotationManager` has an associated `*InteractionDelegate` delegate which can be set in order to get called back on tap events.
- Every `*AnnotationManager` exposes a set of common layer properties that will apply to **every** annotation in its managed collection.
- The `annotations` collection exposes a declarative interface for adding/removing annotations.

#### Annotations
- 4 types: `PointAnnotation`, `PolylineAnnotation`, `PolygonAnnotation`, `CircleAnnotation`.
- Each annotation exposes some customizable properties depending on the layer that backs it.
- Each annotation manages a `Turf.Feature` that is eventually injected into the geojson source of its corresponding manager.

### User impact (optional)

**Creating some point annotations**
```swift
self.pointAnnotationManager = self.mapView.annotations.makePointAnnotationManager()
var p1 = PointAnnotation(point: .init(.init(latitude: 0, longitude: 0)))
p1.image = PointAnnotation.NamedImage(image: UIImage(named: "star")!, name: "star")

var p2 = PointAnnotation(point: .init(.init(latitude: 10, longitude: 10)))
p2.image = PointAnnotation.NamedImage(image: UIImage(named: "star")!, name: "star")

var p3 = PointAnnotation(point: .init(.init(latitude: -10, longitude: -10)))
p3.image = PointAnnotation.NamedImage(image: UIImage(named: "star")!, name: "star")

var p4 = PointAnnotation(point: .init(.init(latitude: -10, longitude: 0)))
p4.image = PointAnnotation.NamedImage(image: UIImage(named: "triangle")!, name: "triangle")

self.pointAnnotationManager?.annotations = [p1, p2, p3, p4]
```

**Creating some line annotations**
```swift
self.lineAnnotationManager = self.mapView.annotations.makePolylineAnnotationManager()
self.lineAnnotationManager?.delegate = self
var lineAnnotation1 = PolylineAnnotation(line: .init(lineCoordinates))
lineAnnotation1.lineColor = .init(color: .red)
lineAnnotation1.lineWidth = 50.0
lineAnnotation1.lineOpacity = 1.0

var lineAnnotation2 = PolylineAnnotation(line: .init([.init(latitude: 0, longitude: 0), .init(latitude: -20, longitude: -20)]))
lineAnnotation2.lineColor = .init(color: .blue)
lineAnnotation2.lineWidth = 10.0
lineAnnotation2.lineOpacity = 0.5

var lineAnnotation3 = PolylineAnnotation(line: .init([.init(latitude: 0, longitude: 0), .init(latitude: 55, longitude: 55)]))
lineAnnotation3.lineColor = .init(color: .orange)
lineAnnotation3.lineWidth = 3
lineAnnotation3.lineOpacity = 1.0
lineAnnotation3.lineBlur = 2

self.lineAnnotationManager?.annotations = [lineAnnotation1, lineAnnotation2, lineAnnotation3]
```

### CODE REVIEW

File: Apps/DebugApp/DebugApp/DebugViewController.swift [hunk 1]
```diff
@@ -83,6 +87,77 @@ public class DebugViewController: UIViewController {
          */
         mapView.mapboxMap.onNext(.mapLoaded) { (event) in
             print("The map has finished loading... Event = \(event)")
+
```
Comment (nishant-karajgikar) [inline Apps/DebugApp/DebugApp/DebugViewController.swift:90]: Will be getting rid of this prior to merge.

File: Sources/MapboxMaps/Annotations/Generated/CircleAnnotationManager.swift [hunk 2]
```diff
@@ -0,0 +1,249 @@
...
+        tapGestureRecognizer.numberOfTapsRequired = 1
+        tapGestureRecognizer.numberOfTouchesRequired = 1
+        view?.addGestureRecognizer(tapGestureRecognizer)
+        tapRecognizer = tapGestureRecognizer
+    }
+    
+    @objc internal func handleTap(_ tap: UITapGestureRecognizer) {
+        let options = RenderedQueryOptions(layerIds: [layerId], filter: nil)
+        mapFeatureQueryable?.queryRenderedFeatures(
+            at: tap.location(in: view),
```
Comment (julianrex) [inline Sources/MapboxMaps/Annotations/Generated/CircleAnnotationManager.swift:204]: We should use a rect here that takes finger size into account, for example at the end of the screen.

File: Sources/MapboxMaps/Annotations/Generated/CircleAnnotationManager.swift [hunk 3]
```diff
@@ -0,0 +1,249 @@
...
+    internal func setupTapRecognizer() {
+        let tapGestureRecognizer = UITapGestureRecognizer(target: self, action: #selector(handleTap(_:)))
+        tapGestureRecognizer.numberOfTapsRequired = 1
+        tapGestureRecognizer.numberOfTouchesRequired = 1
+        view?.addGestureRecognizer(tapGestureRecognizer)
+        tapRecognizer = tapGestureRecognizer
+    }
+    
+    @objc internal func handleTap(_ tap: UITapGestureRecognizer) {
+        let options = RenderedQueryOptions(layerIds: [layerId], filter: nil)
```
Comment (julianrex) [inline Sources/MapboxMaps/Annotations/Generated/CircleAnnotationManager.swift:202]: We previously talked about this being on the orchestrator, but that seems like an optimization that we can address later.

**Review by julianrex — APPROVED** [review]
Looks good to me. 

Discussed doc and selection changes in person.

File: Sources/MapboxMaps/Annotations/Generated/CircleAnnotationManager.swift (hunk 2 above)
Comment (nishant-karajgikar) [reply Sources/MapboxMaps/Annotations/Generated/CircleAnnotationManager.swift:204]: Yeah good point.. will address in a subsequent pr.

File: Sources/MapboxMaps/Annotations/Generated/CircleAnnotationManager.swift (hunk 3 above)
Comment (nishant-karajgikar) [reply Sources/MapboxMaps/Annotations/Generated/CircleAnnotationManager.swift:202]: let's talk about this.. Not sure if it's something we want anymore.

**Pull request merged**